src/tigerbeetle/main.zig

```diff
@@ -446,6 +447,92 @@ const Command = struct {
             watchdog.detach();
         }
 
+        // Try to lock all memory in the process to avoid the kernel swapping pages to disk and
+        // potentially introducting undetectable disk corruption into memory.
+        // This is a best-effort attempt not a hard rule (as it may not cover all memory edge cases)
+        // so warn on error to notify the operator that they should disable swap if possible.
+        const lock_mem_err = "Unable to lock pages in memory ({s}) " ++
+            " - swap paging may trigger undetectable IO to disk";
+        switch (builtin.os.tag) {
+            .linux => {
+                const MCL_CURRENT = 1; // Lock all currently mapped pages.
+                const MCL_ONFAULT = 3; // Lock all pages faulted in (i.e. stack space).
+                const rc = os.linux.syscall1(.mlockall, MCL_CURRENT | MCL_ONFAULT);
+                switch (os.linux.E.init(rc)) {
+                    .AGAIN => log.warn(lock_mem_err, .{"some addresses could not be locked"}),
+                    .NOMEM => log.warn(lock_mem_err, .{"total memory would exceed RLIMIT_MEMLOCK"}),
+                    .PERM => log.warn(lock_mem_err, .{"not enough privileges to lock memory"}),
+                    .INVAL => unreachable, // MCL_ONFAULT specified with MCL_CURRENT.
+                    else => |err| return stdx.unexpected_errno("mlockall", err),
+                }
+            },
+            .macos => {
+                // macOS has mlock() but not mlockall().
+            },
+            .windows => {
```

Comment:
Does the Windows case need a call to [VirtualLock](https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtuallock)?

Reply:
`VirtualLock` takes in a range of memory, acting closer to `mlock` than `mlockall`. We could technically have our own `mmap/VirtualAlloc` bump allocator for main.start then mlock that allocated memory range. But this would exclude things like globals and stack so not as thorough as `SetProcessWorkingSetSize`.